### ISSUE TITLE
Develop method for doing null and exact filtering on fields with Min/Max filter input

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#246]
The null and exact filtering is working for fields that just have one filter input box. But there are lots of fields that users will want to do this filtering on that are numeric and therefore have a Min/Max filter input, which currently doesn't work for the null or exact matching. 

Here is a screen shot with an example -- Gross Floor Area is a numeric value and the filtering has two fields, Min and Max. So currently it is not possible to filter for all the records that have no data (null) for this field. 
![image](https://cloud.githubusercontent.com/assets/6314950/7662441/55c938e2-fb11-11e4-868b-707aeeea3a31.png)

**Comment by nicholasserra** [comment]
This is probably just an edge case that I need to wire in.

**Comment by RDmitchell** [comment]
OK, great, thanks !

**Comment by nicholasserra** [comment]
How do you see this being visually represented? Just put the `""` null filter in one of the inputs? Or both? Not sure.

**Comment by RDmitchell** [comment]
Right, I am not sure either. Maybe "" in either of the input boxes?

**Comment by RDmitchell** [comment]
And can we add the !"" as another case to be covered in this issue?

**Comment by nicholasserra** [comment]
Sounds good, and yes

**Comment by nicholasserra** [comment]
Are there also Min Max fields on date fields? The code leads me to believe there is. 

There are some FE changes that need to be made to support this, specifically with the input types and angular's validation on them. This may end up being a much larger task than originally thought because we'll need to add more sanitation to either the FE or the BE.

Currently, the input for those numeric min max fields in a `type="number"`, which won't do any ajax calls unless the value is numeric. I need to change that to allow any text. 

I then need to filter out invalid values from the payload, making sure only numbers or `""` special filters are passed through. This is where i'm wondering if I also need to handle a date case.

**Comment by RDmitchell** [comment]
Yes, there are min/max input boxes for date fields, ie, the ones that are known as date fields because they are in the Building SnapShot table.

![image](https://cloud.githubusercontent.com/assets/6314950/8921444/7e4bc020-348a-11e5-87db-2514e5d1b414.png)

**Comment by RDmitchell** [comment]
If you think this is going to be a big task, maybe you should try to do an estimate on the level of effort before starting on it, and then we can assess if we want to tackle it now.

**Comment by nicholasserra** [comment]
WIll do. Do you have a sample csv with date fields in it for testing? Or you can just paste a row in here that has a date field. I just need to know the format. I have this all working for numeric fields, but I need to make sure date fields work too.

**Comment by nicholasserra** [comment]
Ping @RDmitchell on that csv. This may have been lost in the shuffle today. Thanks!

**Comment by RDmitchell** [comment]
Here is a sample data set with two date fields, Generation Date and Release Date

https://drive.google.com/a/lbl.gov/file/d/0B3fTKpZ9Dx7LbFZnUHJyOG16N3c/view?usp=sharing

Import it as a "Portfolio Manager" file

![image](https://cloud.githubusercontent.com/assets/6314950/8946183/4e9daa82-3543-11e5-8797-f38f5bf741af.png)

Use this for the mapping

![image](https://cloud.githubusercontent.com/assets/6314950/8946391/a33fa648-3544-11e5-8d43-07284df08a7a.png)

Then once the mapping is saved, and the data matched, you should see this in the Building List view

![image](https://cloud.githubusercontent.com/assets/6314950/8946431/08ff7a4e-3545-11e5-9a32-4456d5b80cfc.png)

**Comment by nicholasserra** [comment]
Awesome, thanks. What's the format for those fields? I was trying the same format as in the column but not getting anything to filter.

**Comment by nicholasserra** [comment]
It looks like the version of angular we're on doesn't even support the date inputs. So i'm wondering if this is broken in general.

**Comment by danielmcquillen** [comment]
Just a thought: there is an the Angularized-version of Bootstrap available here: https://angular-ui.github.io/bootstrap/  It's pretty easy to install via bower.

This makes the Bootstrap controls easier to use within the Angular controller-view setup. And a quick check shows it has a datepicker. 

However, adding something like this (if it is makes sense) would probably entail a bit of refactoring, so it might just be something to consider in Q4. Perhaps it could be used right away to get the datepicker functionality...I don't know if it would collide with the conventional Bootstrap controls.

**Comment by RDmitchell** [comment]
Testing on seedtest
Loaded the sample data for this issue
In the Mapping screen (before saving mapping) It seems like the filtering is working 

![image](https://cloud.githubusercontent.com/assets/6314950/9043153/3edd6860-39c8-11e5-8719-444e9ecf5a6d.png)

But then when I try to reset the filters (using the "x") so that I can see all 8 records in the example file, the program doesn't display them (unless I didn't wait long enough).

Then I save the mapping and "View my buildings".
![image](https://cloud.githubusercontent.com/assets/6314950/9043278/3195dca4-39c9-11e5-9fce-3f8648626c97.png)

Filtering to show Generation Dates between 3/28/2013 and 4/15/2013, it seems to be working.
![image](https://cloud.githubusercontent.com/assets/6314950/9043241/eec74a5c-39c8-11e5-847b-1c5685e8d08d.png)

Same issue to reset to no filtering -- using the "x" to the right of the date in each filter box, it clears the dates I entered but the data doesn't refresh. I can make all 8 records redisplay by refreshing at the browser level.

So it seems like the date filtering for min/max is working, with a few clean up issues.

**Comment by nicholasserra** [comment]
Got it. I was having issues because I was not using the datepicker, and it wasn't triggering the update.

**Comment by nicholasserra** [comment]
So, if we want to allow `""` and "!""` filters t go in these inputs, it would require changing the `type="date"` and `type="number"` attrs on these inputs to `type="text"`.

I have this kind of working by adding custom regex to the controllers for these fields, but I just noticed the other side effect is that i'm losing the date and numeric helpers (arrows and calendar) on these fields. Those can probably be added back in with some kind of class on the FE, but I haven't even begun to dig into that. Seems that angular ties a lot of stuff into these input types.

Robin, let me know if you want me to proceed with this, or if you feel like this is becoming too large in scope. Looks like it's going to require both FE and BE changes to get this working nicely.

**Comment by RDmitchell** [comment]
Hmmm... it sounds like we probably don't want to lose those types. 

So this means we can't figure out which date and number fields have no data in them without changing those fields to text fields? That doesn't sound like the right approach. 

Maybe we punt on this until the Fall, and try to find a better method for doing this, such as potentially what Daniel suggests, or investigate other solutions then?

**Comment by nicholasserra** [comment]
Well, angular won't accept `""` in those fields if they're set as type number or date. And overriding the regex for it doesn't seem to work. So it seems the approach will be to change those input types, set up custom regex for the inputs, and also set custom classes on those inputs to trigger the datepicker and number picker widgets. It's doable, but I haven't gone down that rabbit hole yet because I wasn't sure how much more time we want to spend on making this work.

**Comment by RDmitchell** [comment]
Before we went down that path, I would want to get a sense of the level of effort. If you can try to figure out how much work it would be, that would be a useful piece of information. But don't implement anything until we decide whether it is worth putting the resources on it that you determine. Thanks.

**Comment by nicholasserra** [comment]
Assigning back to @RDmitchell for more planning.

In my opinion, it's going to take some hacky stuff on the FE with angular, mainly because of those input type restrictions. If I were going at this again, i'd say I could easily burn a week on this.

**Comment by RDmitchell** [comment]
Have tested, am closing
